File: python_port/Basin_Pendulum/Pen_stable_check.jl
using DifferentialEquations, Plots, Statistics

function is_stable(θ₀, ω₀)
    function pendulum!(du, u, p, t)
        a, T , K = p
        du[1] = u[2]                  # u[1] is θ
        du[2] = -a*u[2] + T - K*sin(u[1])    #u[2] is ω
    end

    a = 0.1
    K = 1
    T = 0.5

    u₀ = [θ₀, ω₀] # initial state
    tspan = (0.0, 100.0) # time span for simulation
    p = [a, T , K] # parameters

    prob = ODEProblem(pendulum!, u₀, tspan, p)
    sol = solve(prob, Tsit5(), reltol=1e-5, abstol=1e-5)

    θ₁ = asin(T / K)
    adjusted_angle = sol[1, :] .- θ₁

    last_angle = adjusted_angle[end-10:end-1] - adjusted_angle[end-9:end]
    last_speed = sol[2, end-10:end-1]  - sol[2, end-9:end]
    sq_last = sum(last_angle.^2) + sum(last_speed.^2)

    return sq_last < 1
end

File: python_port/Basin_Pendulum/Pen_stable_check.py
import numpy as np
from scipy.integrate import solve_ivp

def pendulum(t, u, a, T, K):
    # u[0] is θ, u[1] is ω
    du = np.zeros_like(u)
    du[0] = u[1]
    du[1] = -a * u[1] + T - K * np.sin(u[0])
    return du

def is_stable(theta_0, omega_0):
    a = 0.1
    K = 1
    T = 0.5

    u_0 = [theta_0, omega_0]  # initial state
    t_span = (0.0, 100.0)  # time span for simulation
    t_eval = np.linspace(t_span[0], t_span[1], 1000)  # evaluate at 1000 points for resolution
    params = (a, T, K)  # parameters

    # Solve ODE
    sol = solve_ivp(pendulum, t_span, u_0, args=params, t_eval=t_eval, rtol=1e-5, atol=1e-5)

    theta_1 = np.arcsin(T / K)
    adjusted_angle = sol.y[0] - theta_1

    last_angle = adjusted_angle[-10:-1] - adjusted_angle[-9:]
    last_speed = sol.y[1][-10:-1] - sol.y[1][-9:]
    sq_last = np.sum(last_angle**2) + np.sum(last_speed**2)

    return sq_last < 1

# # Example usage
# stable = is_stable(np.pi/4, 0)  # Example initial conditions
# print("Is stable:", stable)
